File: src/arm_controller/example/example_dual_arm.cpp
#include "arm_controller/arm_controller_api.hpp"
#include "controller/movej/movej_ipc_interface.hpp"
#include "controller/movel/movel_ipc_interface.hpp"
#include "controller/movec/movec_ipc_interface.hpp"
#include <iostream>
#include <thread>
#include <chrono>

using namespace arm_controller;

int main() {
    std::cout << "===============================================================\n"
              << "ARM Controller IPC 演示\n"
              << "===============================================================\n\n";

    // 初始化 IPC
    if (!IPCLifecycle::initialize()) {
        std::cerr << "❌ 初始化失败\n";
        return 1;
    }
    std::cout << "✅ 初始化成功\n\n";

    // 创建接口实例
    movej::MoveJIPCInterface movej;
    movel::MoveLIPCInterface movel;
    movec::MoveCIPCInterface movec;

    // 1. MoveJ 演示 - left_arm 先到达关节目标
    std::cout << "========== MoveJ 演示 ==========\n";
    std::cout << "发送 MoveJ 命令 -> left_arm ...\n";
    if (!movej.execute({0.0, -0.5236, -0.7854, 0.0, 0.5236, 0.0}, "left_arm")) {
        std::cerr << "❌ 失败: " << movej.getLastError() << "\n";
    } else {
        std::cout << "✅ 已入队\n";
    }
    std::cout << "\n";

    // 2. MoveL 演示 - left_arm 从关节空间目标直线到笛卡尔目标 (0.19, -0.5, 0.63)
    std::cout << "========== MoveL 演示 ==========\n";
    std::cout << "发送 MoveL 命令 -> left_arm ...\n";
    if (!movel.execute(0.19, -0.5, 0.63, -0.4546, 0.4546, -0.5417, 0.5417, "left_arm")) {
        std::cerr << "❌ 失败: " << movel.getLastError() << "\n";
    } else {
        std::cout << "✅ 已入队\n";
    }
    std::cout << "\n";

    // 3. MoveJ 演示
    std::cout << "========== MoveJ 演示 ==========\n";
    std::cout << "发送 MoveJ 命令 -> right_arm ...\n";
    if (!movej.execute({0.0, -0.5236, -0.7854, 0.0, 0.5236, 0.0}, "right_arm")) {
        std::cerr << "❌ 失败: " << movej.getLastError() << "\n";
    } else {
        std::cout << "✅ 已入队\n";
    }
    std::cout << "\n";

    // 4. MoveL 演示
    std::cout << "========== MoveL 演示 ==========\n";
    std::cout << "发送 MoveL 命令 -> right_arm ...\n";
    if (!movel.execute(0.19, 0.5, 0.63, -0.4546, 0.4546, -0.5417, 0.5417, "right_arm")) {
        std::cerr << "❌ 失败: " << movel.getLastError() << "\n";
    } else {
        std::cout << "✅ 已入队\n";
    }
    std::cout << "\n";

    // 3. MoveC 演示 - 从 MoveL 目标位置出发的小范围弧形轨迹
    std::cout << "========== MoveC 演示 ==========\n";
    std::cout << "发送 MoveC 命令 (从 MoveL 终点出发)... -> left_arm \n";
    // 起点：MoveL 目标位置 (0.19, -0.5, 0.63)
    // 中间点：向内下方偏移 (+4cm, 0cm, -5cm) = (0.23, -0.5, 0.58)
    // 目标点：继续偏移 (+6cm, 0cm, -8cm) = (0.25, -0.5, 0.55)
    // 形成一个小范围的圆弧，方便规划
    if (!movec.execute({
        0.2300, -0.5000, 0.5800, -0.4546, 0.4546, -0.5417, 0.5417,  // via_point
        0.2500, -0.5000, 0.5500, -0.4546, 0.4546, -0.5417, 0.5417   // goal_point
    }, "left_arm")) {
        std::cerr << "❌ 失败: " << movec.getLastError() << "\n";
    } else {
        std::cout << "✅ 已入队\n";
    }
    std::cout << "\n";

    // 关闭 IPC
    std::cout << "关闭 IPC...\n";
    IPCLifecycle::shutdown();

    std::cout << "===============================================================\n"
              << "演示结束！\n"
              << "===============================================================\n";
    return 0;
}

File: src/arm_controller/example/example_dual_arm.py
#!/usr/bin/env python3

import argparse
import os
import sys
import time


def try_import_from_workspace():
    current_dir = os.path.dirname(os.path.abspath(__file__))
    candidates = [
        os.path.join(current_dir, "..", "..", "install", "arm_controller", "lib", "arm_controller"),
        os.path.join(current_dir, "..", "..", "..", "install", "arm_controller", "lib", "arm_controller"),
    ]

    for candidate in candidates:
        candidate = os.path.abspath(candidate)
        if os.path.isdir(candidate) and candidate not in sys.path:
            sys.path.insert(0, candidate)

    import arm_controller_py as ac
    return ac


def parse_args():
    parser = argparse.ArgumentParser(
        description="ARM Controller MoveJ IPC Python demo"
    )
    parser.add_argument(
        "--mapping",
        choices=["left_arm", "right_arm", "both"],
        default="left_arm",
        help="Target arm mapping",
    )
    parser.add_argument(
        "--positions",
        nargs=6,
        type=float,
        default=[0, 0.52, 1.55, 1.55, 1.55, 2.6],
        metavar=("J1", "J2", "J3", "J4", "J5", "J6"),
        help="Target joint positions in radians",
    )
    parser.add_argument(
        "--wait",
        action="store_true",
        help="Wait until the controller reports SUCCESS or FAILED",
    )
    parser.add_argument(
        "--timeout",
        type=float,
        default=15.0,
        help="Wait timeout in seconds when --wait is enabled",
    )
    parser.add_argument(
        "--poll-interval",
        type=float,
        default=0.1,
        help="Polling interval in seconds when --wait is enabled",
    )
    return parser.parse_args()


def wait_for_completion(ac, movej, mapping, timeout_s, poll_interval_s):
    deadline = time.monotonic() + timeout_s

    while time.monotonic() < deadline:
        state = movej.get_execution_state(mapping)
        if state == ac.ExecutionState.SUCCESS:
            print(f"[{mapping}] execution succeeded")
            return True
        if state == ac.ExecutionState.FAILED:
            print(f"[{mapping}] execution failed: {movej.get_last_error()}")
            return False
        time.sleep(poll_interval_s)

    print(f"[{mapping}] wait timeout, last_state={movej.get_execution_state(mapping)}")
    return False


def send_movej(ac, mapping, positions, wait, timeout_s, poll_interval_s):
    movej = ac.MoveJIPCInterface()

    print(f"Sending MoveJ command -> {mapping}: {positions}")
    if not movej.execute(positions, mapping):
        print(f"[{mapping}] failed: {movej.get_last_error()}")
        return False

    print(
        f"[{mapping}] queued, mode={movej.get_current_mode(mapping)}, "
        f"state={movej.get_execution_state(mapping)}"
    )

    if wait:
        return wait_for_completion(ac, movej, mapping, timeout_s, poll_interval_s)
    return True


def main():
    args = parse_args()
    ac = try_import_from_workspace()

    print("===============================================================")
    print("ARM Controller MoveJ Python Demo")
    print("===============================================================\n")

    if not ac.IPCLifecycle.initialize():
        print("IPC initialize failed")
        raise SystemExit(1)

    mappings = ["left_arm", "right_arm"] if args.mapping == "both" else [args.mapping]

    success = True
    for mapping in mappings:
        ok = send_movej(
            ac,
            mapping,
            args.positions,
            args.wait,
            args.timeout,
            args.poll_interval,
        )
        success = success and ok

    ac.IPCLifecycle.shutdown()

    if not success:
        raise SystemExit(1)


if __name__ == "__main__":
    main()
